### ISSUE TITLE
Notes are not grouped by category

### ISSUE THREAD

**Issue by mrzapp** [stefan-niedermann/nextcloud-notes#902]
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are interested into the same feature.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

**Describe the bug**
No matter which sorting method is chosen (chronological or alphabetical) the notes are not grouped into their respective folders. Their folders are only displayed above them.

**To Reproduce**
1. Open the sidebar
2. Pick a category
3. Notes are displayed "flat" instead of categorised

**Expected behavior**
The notes should be grouped into the categories they belong to, like in the web app.

**Smartphone (please complete the following information):**
 - Nextcloud Notes-Version (android app): 2.16.2
 - F-Droid or Play Store: F-Droid
 - Android-Version: 10
 - Device: Moto G5S Plus

**Server**
 - Nextcloud version: 19
 - Nextcloud Notes version (server app): (latest)

**Comment by stefan-niedermann** [comment]
Sorry, but i have no idea what you are trying to tell me 😄

Categorization and sorting are in no way related. You can choose whether the notes in one category are sorted by modified date or by alphabet. For example you can set all Notes in the category "Music" are sorted alphabetically and all Notes which are displayed in the category "ToDo" are sorted by last modified.

Please provide exact steps to reproduce, describe what exactly you are clicking (each step) and what you are expecting.

Ideally with screenshots 😉

**Comment by stefan-niedermann** [comment]
Do you mean that in the categories there are no **sections** between e.g. "A", and " B" like on "All notes" view?

**Comment by mrzapp** [comment]
Sorry for the confusing description 😅 Maybe this clears it up.

This is the web app (sectioned in categories):
![Screenshot_20200627-062958](https://user-images.githubusercontent.com/2747428/85914607-c8d3de00-b83f-11ea-8c0d-5d048ae0c687.png)

This is the Android app (ignoring categories):
![Screenshot_20200627-062841](https://user-images.githubusercontent.com/2747428/85914620-ebfe8d80-b83f-11ea-943a-4a10ac3ea9b3.png)

I get how the alphabetical or chronological "flat" views can be handy, but as of right now, there seems to be no way to explore notes by category, which makes the feature kind of meaningless. I have categories because I don't remember what my notes are called or when I last updated them.

Thank you for this amazing app, though, I use it very frequently 😁 If you'd prefer, I can look into making a pull request fixing this.

**Comment by mrzapp** [comment]
IMO the sectioned categories should be in the alphabetical view, but it might also be something worth implementing as a third view option, so the "flat" view will still be available to those who like it.

**Comment by korelstar** [comment]
I think this is a bug, since the alphabetical view was grouped by (sub) categories before (exactly like you want). Maybe, the bug was introduced while adding the sorting option?

I didn't noticed before, since I just upgraded to that version.

**Comment by korelstar** [comment]
"All notes" view is fine. But in category view, there is another issue: sorting by date should include date/time headers (like in "All notes" view) instead of category headers.

**Comment by stefan-niedermann** [comment]
Thank you for the explanation 🙂 

- yes this is definitely a bug
- no we don't need a third option for this

> I can look into making a pull request fixing this.

- This would be awesome 🚀 Let me know if I can provide you any help @mrzapp 

cc optionally @Isaac-Graham @WingsUpete @Hui-Ouyang16 and @IF-ACT who originally implemented the sorting feature in case they are interested in this bug

**Comment by IF-ACT** [comment]
I guess the sorting method should be modified like this:
- In alphabetical view, sort notes firstly by subcategories then alphabetical order.
- In last-modified view, show time slotter instead of subcategories.

I never use subcategory feature so I didn't notice this bug...

**Comment by stefan-niedermann** [comment]
I think also in last-modified view the notes should be sorted by subcategories first and then by last-modified (as it was before we had sorting).

When one is in a category view, the section headers should always be the subcategories (if available)

**Comment by IF-ACT** [comment]
In `NotesDatabase.java` line 361, change
```java
if (sortingMethod != null) {
    order = key_favorite + " DESC," + sortingMethod.getSorder();
}
```
to
```java
if (sortingMethod != null) {
    if (category != null) {
        order += "," + key_favorite + " DESC," + sortingMethod.getSorder();
    } else {
        order = key_favorite + " DESC," + sortingMethod.getSorder();
    }
}
```
may fix this (I haven't test)

**Comment by stefan-niedermann** [comment]
That's not all unfortunately. While this will restore the grouping by subcategories, the notes within the subcategories are not sorted correctly.

**Comment by stefan-niedermann** [comment]
My bad. [The proposed fix](https://github.com/stefan-niedermann/nextcloud-notes/issues/902#issuecomment-650510926) works very well and will soon be published in `v2.16.4`. :slightly_smiling_face: Thanks again @IF-ACT
